### PR TITLE
fix(wlog/watcher_test.go): make TestRun_AvoidNotifyWhenBehind more resilient

### PR DESCRIPTION
**Pull request by machine424**
https://github.com/prometheus/prometheus/pull/14446#pullrequestreview-2166891678 follow-up

### CODE REVIEW

File: tsdb/wlog/watcher_test.go
Comment (bboreham) [inline tsdb/wlog/watcher_test.go:771]: It's not allowed to call t.FailNow on a background goroutine.
Perhaps use golang.org/x/sync/errgroup and return the error?

Comment (cstyan) [reply tsdb/wlog/watcher_test.go:778]: `require.NoError` calls `t.FailNow` under the hood, so I think Bryan's comment here is still relevant

**Review by cstyan — COMMENTED** [review]
LGTM besides the one open comment, but I guess I need to audit all of the remote write tests for failure cases in goroutines now!

File: tsdb/wlog/watcher_test.go
Comment (machine424) [reply tsdb/wlog/watcher_test.go:778]: Yes, I'm aware of the "issue," but I've always ignored it and have never encountered a false "test passed" because of it. As long as the test fails when necessary, I'm not too concerned. That said, I'm okay with implementing a cleaner solution.

We already use this in other places (Go really just needs to support it; it's simpler and more natural).

That's why `testifylint`'`go-require` that help catch some of them is disabled: https://github.com/prometheus/prometheus/blob/349068ad3ed81d32aa78b8a66da9a6fbec1ac333/.golangci.yml#L152-L155C9

**Comment by machine424** [comment]
> LGTM besides the one open comment, but I guess I need to audit all of the remote write tests for failure cases in goroutines now!

As of now, enabling testifylint'go-require yields:


```
discovery/xds/xds_test.go:51:3: go-require: do not use require in http handlers (testifylint)
		require.Equal(t, "application/json", r.Header.Get("Content-Type"))
		^
discovery/xds/xds_test.go:52:3: go-require: do not use require in http handlers (testifylint)
		require.Equal(t, "application/json", r.Header.Get("Accept"))
		^
discovery/xds/xds_test.go:59:3: go-require: do not use require in http handlers (testifylint)
		require.NotEmpty(t, body)
		^
discovery/xds/xds_test.go:60:3: go-require: do not use require in http handlers (testifylint)
		require.NoError(t, err)
		^
discovery/xds/xds_test.go:65:3: go-require: do not use require in http handlers (testifylint)
		require.NoError(t, err)
		^
discovery/xds/xds_test.go:80:3: go-require: do not use require in http handlers (testifylint)
		require.NoError(t, err)
		^
discovery/xds/xds_test.go:83:3: go-require: do not use require in http handlers (testifylint)
		require.NoError(t, err)
		^
notifier/notifier_test.go:863:3: go-require: do not use require in http handlers (testifylint)
		require.NoError(t, err)
		^
notifier/notifier_test.go:866:3: go-require: do not use require in http handlers (testifylint)
		require.NoError(t, err)
		^
notifier/notifier_test.go:949:3: go-require: do not use require in http handlers (testifylint)
		require.NoError(t, err)
		^
notifier/notifier_test.go:952:3: go-require: do not use require in http handlers (testifylint)
		require.NoError(t, err)
		^
scrape/scrape_test.go:2348:5: go-require: do not use require in http handlers (testifylint)
				require.Truef(t, strings.Contains(accept, "escaping=allow-utf-8"), "Expected Accept header to allow utf8, got %q", accept)
				^
scrape/scrape_test.go:2351:5: go-require: do not use require in http handlers (testifylint)
				require.True(t, strings.HasPrefix(accept, "application/vnd.google.protobuf;"),
				^
scrape/scrape_test.go:2356:4: go-require: do not use require in http handlers (testifylint)
			require.Equal(t, expectedTimeout, timeout, "Expected scrape timeout header.")
			^
scrape/scrape_test.go:3455:5: go-require: do not use require in http handlers (testifylint)
				require.Equal(t, tc.acceptEncoding, r.Header.Get("Accept-Encoding"), "invalid value of the Accept-Encoding header")
				^
scrape/scrape_test.go:3748:3: go-require: require must only be used in the goroutine running the test function (testifylint)
		require.NoError(t, err)
		^
scrape/target_test.go:151:5: go-require: do not use require in http handlers (testifylint)
				require.Equal(t, expected, received, "Authorization header was not set correctly.")
				^
scrape/target_test.go:172:5: go-require: do not use require in http handlers (testifylint)
				require.Equal(t, expected, received, "Authorization header was not set correctly.")
				^
scrape/target_test.go:192:5: go-require: do not use require in http handlers (testifylint)
				require.True(t, ok, "Basic authorization header was not set correctly.")
				^
scrape/target_test.go:193:5: go-require: do not use require in http handlers (testifylint)
				require.Equal(t, "user", username)
				^
scrape/target_test.go:194:5: go-require: do not use require in http handlers (testifylint)
				require.Equal(t, "password123", password)
				^
storage/remote/client_test.go:181:5: go-require: do not use require in http handlers (testifylint)
				require.Equal(
				^
storage/remote/queue_manager_test.go:540:6: go-require: require must only be used in the goroutine running the test function (testifylint)
					require.True(t, sent, "samples not sent")
					^
storage/remote/storage_test.go:172:4: go-require: require must only be used in the goroutine running the test function (testifylint)
			require.NoError(t, s.ApplyConfig(conf))
			^
tsdb/compact_test.go:1680:8: go-require: require must only be used in the goroutine running the test function (testifylint)
							require.NoError(t, err)
							^
tsdb/compact_test.go:1683:6: go-require: require must only be used in the goroutine running the test function (testifylint)
					require.NoError(t, sparseApp.Commit())
					^
tsdb/compact_test.go:1689:6: go-require: require must only be used in the goroutine running the test function (testifylint)
					require.NoError(t, err)
					^
tsdb/compact_test.go:1691:6: go-require: require must only be used in the goroutine running the test function (testifylint)
					require.NoError(t, err)
					^
tsdb/compact_test.go:1692:6: go-require: require must only be used in the goroutine running the test function (testifylint)
					require.Len(t, sparseULIDs, 1)
					^
tsdb/compact_test.go:1716:9: go-require: require must only be used in the goroutine running the test function (testifylint)
								require.NoError(t, err)
								^
tsdb/compact_test.go:1723:8: go-require: require must only be used in the goroutine running the test function (testifylint)
							require.NoError(t, err)
							^
tsdb/compact_test.go:1729:8: go-require: require must only be used in the goroutine running the test function (testifylint)
							require.NoError(t, err)
							^
tsdb/compact_test.go:1734:6: go-require: require must only be used in the goroutine running the test function (testifylint)
					require.NoError(t, oldApp.Commit())
					^
tsdb/compact_test.go:1740:6: go-require: require must only be used in the goroutine running the test function (testifylint)
					require.NoError(t, err)
					^
tsdb/compact_test.go:1742:6: go-require: require must only be used in the goroutine running the test function (testifylint)
					require.NoError(t, err)
					^
tsdb/compact_test.go:1743:6: go-require: require must only be used in the goroutine running the test function (testifylint)
					require.Len(t, oldULIDs, 1)
					^
tsdb/db_test.go:1396:5: go-require: require must only be used in the goroutine running the test function (testifylint)
				require.NoError(t, db.reloadBlocks())
				^
tsdb/db_test.go:1401:5: go-require: require must only be used in the goroutine running the test function (testifylint)
				require.NoError(t, db.CleanTombstones())
				^
tsdb/db_test.go:2733:5: go-require: require must only be used in the goroutine running the test function (testifylint)
				require.NoError(t, err)
				^
tsdb/db_test.go:2736:4: go-require: require must only be used in the goroutine running the test function (testifylint)
			require.NoError(t, err)
			^
tsdb/db_test.go:3088:5: go-require: require must only be used in the goroutine running the test function (testifylint)
				require.NoError(t, err)
				^
tsdb/db_test.go:3089:5: go-require: require must only be used in the goroutine running the test function (testifylint)
				require.Nil(t, iterable)
				^
tsdb/db_test.go:3090:5: go-require: require must only be used in the goroutine running the test function (testifylint)
				require.Equal(t, chunk.Chunk.Bytes(), chkAct.Bytes())
				^
tsdb/db_test.go:3775:3: go-require: require must only be used in the goroutine running the test function (testifylint)
		require.NoError(t, db.CompactOOOHead(ctx))
		^
tsdb/db_test.go:3776:3: go-require: require must only be used in the goroutine running the test function (testifylint)
		require.Equal(t, float64(1), prom_testutil.ToFloat64(db.Head().metrics.chunksRemoved))
		^
tsdb/db_test.go:3873:3: go-require: require must only be used in the goroutine running the test function (testifylint)
		require.NoError(t, db.CompactOOOHead(ctx))
		^
tsdb/db_test.go:3874:3: go-require: require must only be used in the goroutine running the test function (testifylint)
		require.Equal(t, float64(1), prom_testutil.ToFloat64(db.Head().metrics.chunksRemoved))
		^
tsdb/db_test.go:3964:3: go-require: require must only be used in the goroutine running the test function (testifylint)
		require.NoError(t, db.CompactOOOHead(ctx))
		^
tsdb/db_test.go:3965:3: go-require: require must only be used in the goroutine running the test function (testifylint)
		require.Equal(t, float64(1), prom_testutil.ToFloat64(db.Head().metrics.chunksRemoved))
		^
tsdb/head_test.go:3328:3: go-require: require must only be used in the goroutine running the test function (testifylint)
		require.NoError(t, db.Compact(ctx))
```

I can see some `storage/remote/` there.

File: tsdb/wlog/watcher_test.go
Comment (machine424) [reply tsdb/wlog/watcher_test.go]: done, and adjusted some comments.

**Review by cstyan — APPROVED** [review]
@machine424 I think we should merge this and see if it fixes the flake in CI

**Comment by machine424** [comment]
> @machine424 I think we should merge this and see if it fixes the flake in CI

it was waiting for an approval, then I made it a draft as I think we can do without an extra goroutine, but we can keep that for later.

**Pull request merged**
